docs/integrations/Instructions.md

```diff
@@ -0,0 +1,160 @@
+
+Certain fields may be omitted depending on the network in question.
+
+### In Instructions
+
+  - `origin` (Address): Address from the network of origin which sent funds in.
+  - `target` (Address): The ink! contract to transfer the incoming funds to.
+  - `data`   (Vec<u8>): The data to call the target with.
+
+Networks may automatically provide `origin`. If they do, the instruction may
+still provide `origin`, overriding the automatically provided value. If no
+`origin` is provided, the instruction is dropped.
+
+Upon receiving funds, the respective Serai Asset contract is called, minting the
+appropriate amount of coins, and transferring them to the specified target with
+the attached data.
+
+If the transaction fails, funds are scheduled to be returned to `origin`.
+
+### Out Instructions
+
+  - `destination` (Enum { Native(Address), Serai(Address) }): Address to receive
+funds to.
+  - `data`        (Vec<u8>):                                  The data to call
+the target with.
+
+If the network is Serai, this is a transfer. Else, it's a withdrawal to the
+specified address with the specified data. Asset contracts perform no validation
+on these fields.
+
+### Shorthand
+
+All In Instructions are encoded as Shorthand. Shorthand is an enum which expands
+to an In Instruction.
+
+##### Raw
+
+Raw Shorthand encodes a raw In Instruction with no further processing.
+
+##### Swap
+
+  - `origin`  (Option<Address>): In Instruction's `origin`.
+  - `coin`    (Coin):            Coin to swap funds for.
+  - `minimum` (Amount):          Minimum amount of `coin` to receive.
+  - `out`     (Out Instruction): Final destination for funds.
+
+which expands to:
+
+```
+In Instruction {
+  origin,
+  target: Router,
+  data:   swap(Incoming Asset, out, minimum)
+}
+```
+
+where `swap` is a function which:
+
+  1) Swaps the incoming funds for SRI.
+  2) Swaps the SRI for `coin`.
+  3) Checks the amount of `coin` received is greater than `minimum`.
+  4) Executes `out` with the amount of `coin` received.
+
+For a Bitcoin to Monero swap, Swap Shorthand is expected to generally take:
+
+  - 1 byte to identify as Swap.
+  - 1 byte to not override `origin`.
+  - 1 byte for `coin`.
+  - 4 bytes for `minimum`.
+  - 1 byte for `out`'s `destination` field label.
+  - 1 byte for `out`'s `destination`'s ordinal byte.
+  - 65 bytes for `out`'s `destination`'s address.
+
+Or 74 bytes.
+
+##### Add Liquidity
+
+  - `origin`  (Option<Address>): In Instruction's `origin`.
+  - `minimum` (Amount):  Minimum amount of SRI to receive.
+  - `gas`     (Amount):  Amount of SRI to send to `address` to cover gas in the
+future.
+  - `address` (Address): Account to give the created liquidity tokens.
+
+which expands to:
+
+```
+In Instruction {
+  origin,
+  target: Router,
+  data:   swap_and_add_liquidity(Incoming Asset, address, minimum, gas)
+}
+```
+
+where `swap_and_add_liquidity` is a function which:
+
+  1) Swaps half of the incoming funds for SRI.
```

Comment:
This actually leaves a potentially notable amount of potential liquidity on the table, as it doesn't consider slippage at all. It's just also trivial to calculate, and the amounts aren't notable at all when slippage isn't notable due to the size of the liquidity pool.